### PR TITLE
obs-filters: Fix 'Image Mask/Blend' opacity

### PR DESCRIPTION
**Pull request by Oncorporation**
- Fix 'Opacity' option on GUI:  It is not implemented in the associated filters and is currently a worthless GUI element. 
- Change 'Opacity' GUI to slider
- Implement Opacity in all 'Image Mask/Blend' filters
- Add 'Background Removal Mask' filter

Long standing issue also present in SL OBS fork.

### CODE REVIEW

**Comment by Oncorporation** [comment]
Tried to squash, but it is just adding useless commits.

**Comment by pkviet** [comment]
do an interactive rebase; ex:  git rebase -i HEAD~12 (since there are 12 commits); in the editor, change pick to f for the commits you want to squash into previous, reorder, etc

**Comment by Oncorporation** [comment]
In my history tree, it only shows 4b3a81d4 and 1e017cee, ran the rebase with fixups.

**Comment by pkviet** [comment]
> In my history tree, it only shows [4b3a81d](https://github.com/obsproject/obs-studio/commit/4b3a81d462b62d49ca3a466895fc499b8fe0529c) and [1e017ce](https://github.com/obsproject/obs-studio/commit/1e017ceed3634b9a1e4b141398a24a7b37df3d2c), ran the rebase with fixups.

nope, you've ended up with 14 commits now.
Here's how your tree in your repo looks:


Try using git on the command-line and mark as fixup ALL the commits except the main ones.

**Comment by Oncorporation** [comment]
That is exactly what I did

**Comment by pkviet** [comment]
well ...

**Comment by pkviet** [comment]
you have to do the interactive rebase starting from your first commit which would be a24d6ec
You'll end up with 2 commits , one for luma, the other for the blend stuff.

**Comment by Oncorporation** [comment]
> you have to do the interactive rebase starting from your first commit which would be [a24d6ec](https://github.com/obsproject/obs-studio/commit/a24d6ecce39e35102b546f712490ecb0d5ca79d3)
> You'll end up with 2 commits , one for luma, the other for the blend stuff.

Thanks, I can see the issue, that is why I have made several attempts to squash. This last attempt was force pushed.

**Comment by Oncorporation** [comment]
Ok, now onto the code!
There are a lot of very complex ways to do background removal with hashes, etc.  

This is the minimum viable method. The background image and "shadow" version are compared by pixel and removed if they meet a threshold.  It is a placeholder for further development.

**Comment by Oncorporation** [comment]
Ok, now onto the code!

### Opacity Broken
The 'Opacity' option is broken for 'Image Mask/Blend' in all forks. It is a three year old issue. The option was put in the GUI and not implemented except to grab a default white color.

All associated masks/blends now use 'Opacity' to adjust how much they are applied to the source image.

### Background Removal Mask
There are many complex methods of background removal using hashes, blurs, samples, etc.

This is the minimum viable method. The background image and "shadow" version are compared by pixel and removed if they meet a threshold ('Opacity'). It is a placeholder for further development.

To use:

- Take a video of just the background.
- Take a frame and use it as the background image. Windows Snipping Tool is sufficient. (%windir%\system32\SnippingTool.exe)

This eliminates differences based upon camera/video settings. The 'Color' option is then used to adjust for lighting.

**Comment by jpark37** [comment]
I tried current OBS a bit, and the current opacity setting can work as-is. I think it requires an image to be loaded first. This line of code feeds the opacity:

color |= (uint32_t)(((double)opacity) * 2.55) << 24;

It's a little dubious that it does a bitwise or with a potentially existing alpha, and it appears selecting a custom color slams alpha to 255 even if you try to enter 0. (Not sure if I analyzed this correctly; I haven't run a debugger.) The feature is implemented in some form though, so a fix should take that into account.

**Comment by SuslikV** [comment]
> ...
> and it appears selecting a custom color slams alpha to 255 even if you try to enter 0.
> ...

@jpark37 To clarify this: https://github.com/obsproject/obs-studio/commit/78d566916bfd9847e8777a031f590106de17aa36

**Comment by Oncorporation** [comment]
In my review `color |= (uint32_t)(((double)opacity) * 2.55) << 24;` says if no color, use opacity and create white for color.

It does not allow you to modify `Opacity` in the shader.

To confirm, just load a mask or blend without the PR fix and try to adjust `Opacity`

**Comment by Xaymar** [comment]
The addition of the background removal filter should be in its own PR.

**Comment by Oncorporation** [comment]
All the mask/blend effects are modified to implement 'Opacity', so modifying/adding 6 instead of 5 related files seem trivial.

**Comment by Oncorporation** [comment]
All the mask/blend effects are modified to implement 'Opacity', so modifying/adding 6 instead of 5 related files seem trivial to me. More so because it is a significant selling point to end users.

**Comment by Xaymar** [comment]
I am not sure if a feature addition should be in the same PR as fixing a problem. For me a PR should be a minimum viable change, that means that the change should be as small and contained as possible - so I would personally split feature addition and problem fixing into two PRs, one depending on the problem fixing. This is something I picked up when working with ffmpeg.

To me, this is not trivial. You are merging a feature addition with a problem/bug fix, which really shouldn't be together, especially in a single commit.

**Comment by SuslikV** [comment]
By the way, the description doesn't says what it fixes. So, what was wrong with `'Opacity' in 'Mask/Blend Filter'`?

**Comment by Oncorporation** [comment]
> By the way, the description doesn't says what it fixes. So, what was wrong with `'Opacity' in 'Mask/Blend Filter'`?

Updated the description to make that more clear.  It is not implemented in the shaders and currently does nothing, except is used to create a default white color from the default 100% Opacity. Without the PR, it is just a useless GUI element.

File: plugins/obs-filters/data/blend_add_filter.effect [hunk 1]
```diff
@@ -37,7 +38,7 @@ float4 PSAddImageRGBA(VertDataOut v_in) : TARGET
 	float4 rgba = image.Sample(textureSampler, v_in.uv) * color;
 
 	float4 targetRGB = target.Sample(textureSampler, v_in.uv2);
-	rgba.rgb = saturate(rgba.rgb + targetRGB.rgb);
+	rgba.rgb = lerp(rgba.rgb, saturate(rgba.rgb + targetRGB.rgb), opacity * 0.01);
```
Comment (Xaymar) [inline plugins/obs-filters/data/blend_add_filter.effect]: Scaling of opacity should be done on the CPU, not GPU.

File: plugins/obs-filters/data/blend_mul_filter.effect [hunk 2]
```diff
@@ -37,7 +38,7 @@ float4 PSMuliplyImageRGBA(VertDataOut v_in) : TARGET
 	float4 rgba = image.Sample(textureSampler, v_in.uv) * color;
 
 	float4 targetRGB = target.Sample(textureSampler, v_in.uv2);
-	rgba.rgb = saturate(rgba.rgb * targetRGB.rgb);
+	rgba.rgb = lerp(rgba.rgb, saturate(rgba.rgb * targetRGB.rgb), opacity * 0.01);
```
Comment (Xaymar) [inline plugins/obs-filters/data/blend_mul_filter.effect]: Scaling of opacity should be done on the CPU, not GPU.

File: plugins/obs-filters/data/blend_mul_filter.effect [hunk 3]
```diff
@@ -3,6 +3,7 @@ uniform texture2d image;
 
 uniform texture2d target;
 uniform float4 color;
+uniform int opacity;
```
Comment (Xaymar) [inline plugins/obs-filters/data/blend_mul_filter.effect]: opacity should be a float.

File: plugins/obs-filters/data/blend_add_filter.effect [hunk 4]
```diff
@@ -3,6 +3,7 @@ uniform texture2d image;
 
 uniform texture2d target;
 uniform float4 color;
+uniform int opacity;
```
Comment (Xaymar) [inline plugins/obs-filters/data/blend_add_filter.effect]: opacity should be a float.

File: plugins/obs-filters/data/blend_sub_filter.effect [hunk 5]
```diff
@@ -3,6 +3,7 @@ uniform texture2d image;
 
 uniform texture2d target;
 uniform float4 color;
+uniform int opacity;
```
Comment (Xaymar) [inline plugins/obs-filters/data/blend_sub_filter.effect]: opacity should be a float.

File: plugins/obs-filters/data/blend_sub_filter.effect [hunk 6]
```diff
@@ -37,7 +38,7 @@ float4 PSSubtractImageRGBA(VertDataOut v_in) : TARGET
 	float4 rgba = image.Sample(textureSampler, v_in.uv) * color;
 
 	float4 targetRGB = target.Sample(textureSampler, v_in.uv2);
-	rgba.rgb = saturate(rgba.rgb - targetRGB.rgb);
+	rgba.rgb = lerp(rgba.rgb, saturate(rgba.rgb - targetRGB.rgb), opacity * 0.01);
```
Comment (Xaymar) [inline plugins/obs-filters/data/blend_sub_filter.effect]: Scaling of opacity should be done on the CPU, not GPU.

File: plugins/obs-filters/data/mask_alpha_filter.effect [hunk 7]
```diff
@@ -37,7 +38,7 @@ float4 PSAlphaMaskRGBA(VertDataOut v_in) : TARGET
 	float4 rgba = image.Sample(textureSampler, v_in.uv) * color;
 
 	float4 targetRGB = target.Sample(textureSampler, v_in.uv2);
-	rgba.a *= targetRGB.a;
+	rgba.a = lerp(rgba.a, rgba.a * targetRGB.a, opacity * 0.01);
```
Comment (Xaymar) [inline plugins/obs-filters/data/mask_alpha_filter.effect]: Scaling of opacity should be done on the CPU, not GPU.

File: plugins/obs-filters/data/mask_alpha_filter.effect [hunk 8]
```diff
@@ -3,6 +3,7 @@ uniform texture2d image;
 
 uniform texture2d target;
 uniform float4 color;
+uniform int opacity;
```
Comment (Xaymar) [inline plugins/obs-filters/data/mask_alpha_filter.effect]: opacity should be a float.

File: plugins/obs-filters/data/mask_background_filter.effect [hunk 9]
```diff
@@ -0,0 +1,64 @@
+uniform float4x4 ViewProj;
```
Comment (Xaymar) [inline plugins/obs-filters/data/mask_background_filter.effect:1]: This file should not be part of this PR. It is not relevant to the fixed problem, and is a feature addition.

File: plugins/obs-filters/data/mask_color_filter.effect [hunk 10]
```diff
@@ -3,6 +3,7 @@ uniform texture2d image;
 
 uniform texture2d target;
 uniform float4 color;
+uniform int opacity;
```
Comment (Xaymar) [inline plugins/obs-filters/data/mask_color_filter.effect]: opacity should be a float.

File: plugins/obs-filters/data/mask_color_filter.effect [hunk 11]
```diff
@@ -37,7 +38,7 @@ float4 PSColorMaskRGBA(VertDataOut v_in) : TARGET
 	float4 rgba = image.Sample(textureSampler, v_in.uv) * color;
 
 	float4 targetRGB = target.Sample(textureSampler, v_in.uv2);
-	rgba.a *= (targetRGB.r + targetRGB.g + targetRGB.b) / 3.0;
+	rgba.a = lerp(rgba.a, rgba.a * ((targetRGB.r + targetRGB.g + targetRGB.b) / 3.0), opacity * 0.01);
```
Comment (Xaymar) [inline plugins/obs-filters/data/mask_color_filter.effect]: Scaling of opacity should be done on the CPU, not GPU.

File: plugins/obs-filters/mask-filter.c [hunk 12]
```diff
@@ -220,6 +225,9 @@ static void mask_filter_render(void *data, gs_effect_t *effect)
 	param = gs_effect_get_param_by_name(filter->effect, "color");
 	gs_effect_set_vec4(param, &filter->color);
 
+	param = gs_effect_get_param_by_name(filter->effect, "opacity");
+	gs_effect_set_int(param, filter->opacity);
```
Comment (Xaymar) [inline plugins/obs-filters/mask-filter.c]: opacity should be a float and scaled on the CPU (divide by 100.0f).

**Comment by SuslikV** [comment]
> In my review `color |= (uint32_t)(((double)opacity) * 2.55) << 24;` says if no color, use opacity and create white for color ( color = opacity * 255,255,255). This is for a default value.

How it works (the "create white")? And what means "no color"?

File: plugins/obs-filters/data/blend_mul_filter.effect (hunk 3 above)
Comment (Oncorporation) [reply plugins/obs-filters/data/blend_mul_filter.effect]: The creator of this plugin created opacity as an integer in the GUI.  Would that enhancement require another PR?

**Comment by jpark37** [comment]
> In my review `color |= (uint32_t)(((double)opacity) * 2.55) << 24;` says if no color, use opacity and create white for color ( color = opacity * 255,255,255). This is for a default value.

It's actually more like color = float4(color.rgb, color.a | opacity). Maybe the initial implementation could assume the high alpha bits were 0 because bitwise or doesn't make sense otherwise.

The Opacity setting in current OBS does something. If you load an image into the filter, don't touch the color picker, and change the Opacity setting, the output image becomes affected.

File: plugins/obs-filters/data/blend_mul_filter.effect (hunk 3 above)
Comment (Xaymar) [reply plugins/obs-filters/data/blend_mul_filter.effect]: As it is part of the bug fix (since opacity is actually used now), it would not.

File: plugins/obs-filters/data/blend_add_filter.effect (hunk 1 above)
Comment (Oncorporation) [reply plugins/obs-filters/data/blend_add_filter.effect]: > > In my review `color |= (uint32_t)(((double)opacity) * 2.55) << 24;` says if no color, use opacity and create white for color ( color = opacity * 255,255,255). This is for a default value.
> 
> It's actually more like color = float4(color.rgb, color.a | opacity). Maybe the initial implementation could assume the high alpha bits were 0 because bitwise or doesn't make sense otherwise.
> 
> The Opacity setting in current OBS does something. If you load an image into the filter, don't touch the color picker, and change the Opacity setting, the output image becomes affected.

As buggy as that is, it is because there is no color so it is created by Opacity. 

That is why I say that it really evaluates to  "if there is no color, then create a default color from Opacity."

However, that is not in any sense Opacity it is a color. It just happens to have a default Opacity of 1, and the bitwise "<<24"  pushes that to a float4 white.

**Comment by SuslikV** [comment]
```
	uint32_t color = (uint32_t)obs_data_get_int(settings, SETTING_COLOR);
	int opacity = (int)obs_data_get_int(settings, SETTING_OPACITY);
```
just an values from settings

**Comment by Oncorporation** [comment]
> > In my review `color |= (uint32_t)(((double)opacity) * 2.55) << 24;` says if no color, use opacity and create white for color ( color = opacity * 255,255,255). This is for a default value.
> 
> It's actually more like color = float4(color.rgb, color.a | opacity). Maybe the initial implementation could assume the high alpha bits were 0 because bitwise or doesn't make sense otherwise.
> 
> The Opacity setting in current OBS does something. If you load an image into the filter, don't touch the color picker, and change the Opacity setting, the output image becomes affected.

As buggy as that is, it is because there is no color so it is created by Opacity.

That is why I say that it really evaluates to "if there is no color, then create a default color from Opacity."

However, that is not in any sense Opacity it is a color. It just happens to have a default Opacity of 1, and the bitwise "<<24" pushes that to a float4 white.

File: plugins/obs-filters/data/blend_mul_filter.effect (hunk 3 above)
Comment (Oncorporation) [reply plugins/obs-filters/data/blend_mul_filter.effect]: Prior to pushing this PR up, I had tried a float, but it made no difference. 

It even made it more troublesome and time consuming. Because, it seems that 15.15% might be better than 15%... but you can't tell the difference so you strain your eyes trying to see an improvement.

If changed, the slider would still go by 1.0 step values to prevent insanity.

**Comment by jpark37** [comment]
I'm not sure you're reading this line of code correctly.

color |= (uint32_t)(((double)opacity) * 2.55) << 24;

rgb = color & 0x00FFFFFF; // 0x00BBGGRR
alpha = color & 0xFF000000; // 0xAA000000
opacity_component = (uint32_t)(((double)opacity) * 2.55) << 24; // 0x??000000
total_alpha = alpha | opacity_component; // 0x??000000
color = rgb | total_alpha; // 0x??BBGGRR

File: plugins/obs-filters/data/blend_sub_filter.effect (hunk 6 above)
Comment (Oncorporation) [reply plugins/obs-filters/data/blend_sub_filter.effect]: Xaymar, I am unclear what you mean here. Are you referring to the `opacity * 0.01` code or are you referring to the lerp operation itself?

Comment (Xaymar) [reply plugins/obs-filters/data/blend_sub_filter.effect]: I am referring to the `* 0.01`, as there is no need to do this per fragment.

**Comment by SuslikV** [comment]
What you can't understand? The OR ( `|=` ) bitwise to **fixed** 255 at color picker will never do lower. Isn't it obvious?

**Comment by jpark37** [comment]
My recollection is that the default color is transparent white (0x00FFFFFF), and that causes Opacity to to work as long as you don't touch the color picker.

File: plugins/obs-filters/mask-filter.c (hunk 12 above)
Comment (Oncorporation) [reply plugins/obs-filters/mask-filter.c]: I favor multiplication over division. However, this does make me believe that "Scaling of opacity should be done on the CPU, not GPU." comments are regarding this.

Making changes.

**Comment by SuslikV** [comment]
@jpark37 for `int` read from the settings it is NULL or 0. But source has its on color.

**Comment by Oncorporation** [comment]
> My recollection is that the default color is transparent white (0x00FFFFFF), and that causes Opacity to to work as long as you don't touch the color picker.

```
static void mask_filter_defaults(obs_data_t *settings)
{
	obs_data_set_default_string(settings, SETTING_TYPE,
			"mask_color_filter.effect");
	obs_data_set_default_int(settings, SETTING_COLOR, 0xFFFFFF);
	obs_data_set_default_int(settings, SETTING_OPACITY, 100.0f);
}
```

File: plugins/obs-filters/data/blend_mul_filter.effect (hunk 3 above)
Comment (Oncorporation) [reply plugins/obs-filters/data/blend_mul_filter.effect]: slider set to step at 0.05... I think everyone will be happy now!

**Comment by jp9000** [comment]
Hi there, thank you for the bug fix PR.  I understand what you were trying to do with this PR, although your fix was a little bit more than necessary.  All that needed to be added was one line to fix opacity/alpha.  Additionally, your lerp calls would have converted the output to premultiplied alpha, which we don't usually work with.

When making a pull requests and also commits, it's also important to not combine unrelated changes:  for example, changing the opacity widgets to sliders should have at least been a separate commit, and one could argue maybe a separate PR.  At least a separate commit though.

Additionally, the background removal filter was unrelated to the PR, and should have been a separate PR.  May I ask for a separate PR for the background removal change?

I'm going to close this PR with my own fix, along with the slider change.  If you wish, feel free to make a new PR for the background removal filter so we can review that separately.

**Comment by Oncorporation** [comment]
> Hi there, thank you for the bug fix PR. I understand what you were trying to do with this PR, although your fix was a little bit more than necessary. All that needed to be added was one line to fix opacity/alpha. Additionally, your lerp calls would have converted the output to premultiplied alpha, which we don't usually work with.
> 
> When making a pull requests and also commits, it's also important to not combine unrelated changes: for example, changing the opacity widgets to sliders should have at least been a separate commit, and one could argue maybe a separate PR. At least a separate commit though.
> 
> Additionally, the background removal filter was unrelated to the PR, and should have been a separate PR. May I ask for a separate PR for the background removal change?
> 
> I'm going to close this PR with my own fix, along with the slider change. If you wish, feel free to make a new PR for the background removal filter so we can review that separately.

The Lerp call only applied to the rgb values of the colors. It did not affect the underlying alpha of the item the filter is applied to. It is the "Opacity" of the FILTER being applied.

I wish you would have discussed this further with me.
